File: CBase60.m
#import <Foundation/Foundation.h>
#import "CBase60.h"

@implementation CBase60

#pragma mark - Alphabet / Base

+ (const char *)alphabet {
    return "123456789ABCDEFGHJKLMNPQRSTUVWXYZIOabcdefghijkmnopqrstuvwxyz";
}

+ (int)base {
    return 60;
}

#pragma mark - Decode Table

+ (const int8_t *)decodeTable {
    static int8_t table[128];
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        for (int i = 0; i < 128; i++) {
            table[i] = -1;
        }
        const char *a = [self alphabet];
        for (int i = 0; i < 60; i++) {
            table[(uint8_t)a[i]] = (int8_t)i;
        }
    });
    return table;
}

#pragma mark - Encode (Data → Base60, CANONICAL)

+ (NSString *)encode:(NSData *)data {
    if (data.length == 0) {
        return @"1";   // 规范：0 的唯一表示
    }

    const uint8_t *input = data.bytes;
    NSUInteger length = data.length;

    int digits[length * 2];
    memset(digits, 0, sizeof(digits));

    int digitCount = 0;

    for (NSUInteger b = 0; b < length; b++) {
        int carry = input[b];
        int i = 0;

        while (i < digitCount) {
            carry += digits[i] << 8;
            digits[i] = carry % [self base];
            carry /= [self base];
            i++;
        }

        while (carry > 0) {
            digits[digitCount++] = carry % [self base];
            carry /= [self base];
        }
    }

    // 数值为 0
    if (digitCount == 0) {
        return @"1";
    }

    NSMutableString *result = [NSMutableString stringWithCapacity:digitCount];
    const char *alpha = [self alphabet];

    for (int i = digitCount - 1; i >= 0; i--) {
        [result appendFormat:@"%c", alpha[digits[i]]];
    }

    return result;
}

#pragma mark - Decode (Base60 → Data, CANONICAL)

+ (NSData *)decode:(NSString *)string {
    if (string.length == 0) {
        return [NSData data];
    }

    NSUInteger length = string.length;

    uint8_t bytes[length];
    memset(bytes, 0, sizeof(bytes));

    int byteCount = 0;
    const int8_t *table = [self decodeTable];

    for (NSUInteger idx = 0; idx < length; idx++) {
        unichar ch = [string characterAtIndex:idx];
        if (ch >= 128) return nil;

        int value = table[ch];
        if (value < 0) return nil;

        int carry = value;
        int i = 0;

        while (i < byteCount) {
            carry += bytes[i] * [self base];
            bytes[i] = (uint8_t)(carry & 0xff);
            carry >>= 8;
            i++;
        }

        while (carry > 0) {
            bytes[byteCount++] = (uint8_t)(carry & 0xff);
            carry >>= 8;
        }
    }

    NSMutableData *result = [NSMutableData dataWithCapacity:byteCount];
    for (int i = byteCount - 1; i >= 0; i--) {
        [result appendBytes:&bytes[i] length:1];
    }

    return result;
}

@end

File: CBase60.swift
import Foundation


//Base60编码
public enum CBase60 {

    // MARK: - Alphabet
    private static let alphabet = Array(
        "123456789ABCDEFGHJKLMNPQRSTUVWXYZIOabcdefghijkmnopqrstuvwxyz"
    )

    private static let base = 60

    // Decode lookup table (ASCII 0–127)
    private static let decodeTable: [Int8] = {
        var table = Array(repeating: Int8(-1), count: 128)
        for (i, c) in alphabet.enumerated() {
            let v = Int8(i)
            let ascii = Int(c.asciiValue!)
            table[ascii] = v
        }
        return table
    }()

    // MARK: - Encode (Data → Base60 String)
    public static func encode(_ data: Data) -> String {
        guard !data.isEmpty else { return "" }

        var digits = [Int](repeating: 0, count: data.count * 2)
        var digitCount = 1

        for byte in data {
            var carry = Int(byte)
            var i = 0
            while i < digitCount {
                carry += digits[i] << 8
                digits[i] = carry % base
                carry /= base
                i += 1
            }
            while carry > 0 {
                digits[digitCount] = carry % base
                digitCount += 1
                carry /= base
            }
        }
        
        // Preserve leading zeros
        var result = ""
        for byte in data where byte == 0 {
            result.append(alphabet[0])
        }

        // Convert digits to characters
        for i in stride(from: digitCount - 1, through: 0, by: -1) {
            result.append(alphabet[digits[i]])
        }

        return result
    }

    // MARK: - Decode (Base60 String → Data)
    public static func decode(_ string: String) -> Data? {
        guard !string.isEmpty else { return Data() }

        var bytes = [UInt8](repeating: 0, count: string.count)
        var byteCount = 1

        for c in string {
            guard let ascii = c.asciiValue, ascii < 128 else { return nil }
            let value = Int(decodeTable[Int(ascii)])
            if value < 0 { return nil }

            var carry = value
            var i = 0
            while i < byteCount {
                carry += Int(bytes[i]) * base
                bytes[i] = UInt8(carry & 0xff)
                carry >>= 8
                i += 1
            }
            while carry > 0 {
                bytes[byteCount] = UInt8(carry & 0xff)
                byteCount += 1
                carry >>= 8
            }
        }
        var result = Data()
        result.reserveCapacity(byteCount)

        for i in stride(from: byteCount - 1, through: 0, by: -1) {
            result.append(bytes[i])
        }
        return result
    }
    
    
}
